README.md
# PCA-DT_FeatureAnalysis
Performs PCA and DT fittings to assess feature importance

 ===========================================
 
 README File for PCA.py and DecisionTree.py codes
 
 Code Author: Tom Burns
 
 email: tom.burns@canada.ca
 
 Date: Feb-24th 2021
 
 VERSION 1.1.0
 
 ===========================================
 
 ### Table of Contents
 
 1. Installing Python and required Modules
 2. Setting up Codes
 3. Running the Codes

 ===========================================
 
 ### 1. Installing Python and Required Modules
 -------------------------------------------
 #### Step 1: Installing Python
 - Download and Install python3.9 or earlier installer from: https://www.python.org/
   
 #### Step 2: Installing Modules
 - to install a module, from the command line type
 
    pip --install MODULE
    
    where MODULE is the name of the module you are trying to install
    
 - List of required modules:
   1) numpy
   2) pandas
   3) matplotlib
   4) sklearn
 
 ==========================================
 
 ### 2. Setting Up Codes
 ------------------------------------------
 To run the codes, simply edit the top variables in both codes:
 
 *INFILE* - defines the name of the input file that contains your descriptors and
          target values. THIS FILE NEEDS TO BE IN CSV FORMAT (comma delimited)
	  
 *TARGET* - The name of the column in your csv file that corresponds to the target
          classifications
	  
 *OMIT*   - A list of columns in your CSV file you do not want considered in the
          analysis (example: Substance Name)
	  
 *SCALE*  - Do you want the descriptors to be scaled using the Standard Scaler?
          (set to True or False, recommended: True)
 *NTREES* - The number of decision trees you would like to run (only in DecisionTree.py)
 
          - if NTREES > 1, a random selection of descriptors will be selected for
		       each tree.
		       
		    - if NTREES = 1, all descriptors will be considered in fitting the tree
		    
          - Default value is 10,000 Trees
	  
 *BOOT*   - Run a boostrap to calculate the 95% CI for the decision tree accuracies
          (Only in DecisionTree.py
          - Default = False
	  
 *BSTEPS* - Controls the number of boostrapping steps performed in 95% CI calculation
 
          - Default = 1000
	 
 ==========================================
 
 ### 3. Running the Codes
 ------------------------------------------
 #### From a directory:
 
 1. Open the folder containing the python code
 
 2. make sure the csv containing your data is located in this folder
 
 3. change the INFILE variable in PCA.py and DecisionTree.py to match the
    name of your csv file containing the data and target values

 4. Double click PCA.py or DecisionTree.py to run the code
 
 
 #### From a command line:
 
 1. To run this code, make sure your csv file containing the data is located in
 the same directory as the code.
 
 2. change the INFILE in PCA.py and DecisionTree.py to match the name of your
    csv file containing the data and target values
 
 3. in a Terminal window, navigate to the directory containing the code
	- if on Windows, I recommend PowerShell

 4. Once in the correct directory, type:
 
		python PCA.py 			or			python DecisionTree.py
 ==========================================


Changelog.txt
 ==============================================================================
 Changelog v1.1.0
 
 General Fixes:
 - Fixed bug in PCA.py that caused a crash due to inconsistent label lengths
 - Fixed bug in DT plotting
 - Fixed some plot labels
 - Reduced tree size and added accruacy output from trees (subplot 133)
 - Added a bootstrap function to DecisionTrees.py to calculate the mean
   and CI in subplot 133
 - Added welcome splash to both codes
 - Added optional boostrap to DT code
 - Added progress bars to DT code
 - Fixed a bug that caused errors with sets containing more than 99 features (PCA.py)
 - Added variable called MAX_X to plot_results function of DecisionTree.py, to limit the
   number of features shown in subplot 121 and 122
 
 Specific Changes:
 
 Changes to PCA.py
 - resized label arrays in plot_results() to reduce size of f_labels to match f_x
 - Changed labels in PCA plot to PC 1&2 from PCA 1&2
 - Fixed an error that started the counting of PC at 0 instead of 1 in subplot 222
 - Added code version object
 - Added welcome splash, with code parameters'
 - removed n_components option to PCA class call to fix but with "too many components"
   seen when n_features > 99
 

 
 Changes to DecisionTree.py:
 - Increased default to 10,000 trees from 1,000
 - Added lines of code to rotate the xlabels by 90 degress
 - reduced max depth to 2
 - Added accuracy calculation (see accuracy function) that calculates Balanced
   Accruacy
 - Added array to record BAs from all DecisionTrees
 - increased output to three plots and added subplot at 133 showing dist of
   balanced accuracies
 - Added and implemented a bootstrap function to add CI to subplot 133
 - Added sys import
 - Added progressbar function
 - Added code version object
 - Added welcome splash, with code parameters
 - Added options to toggle Boostrapping, and control the number of boot steps:
   BOOT and BSTEPS
 - Added variable called MAX_X to plot_results function of DecisionTree.py, to limit the
   number of features shown in subplot 121 and 122
 ==============================================================================
 Changelog v1.0.1
 
 General Fixes:
 - Fixed inconsistent plot formatting
 - Made changes to allow the code to be run by double clicking the icon
   instead of running from the command line

 Specific Changes: 
 
 Changes to PCA.py:
 - added os import
 - Added some code to point the python script towards the full path of the
   input file.
 - modified import_data function
 
 Changes to DecisionTree.py:
 - added os import
 - Added some code to point the python script towards the full path of the
   input file.
 - Changed formatting to the bar graphs, to be consistent with the formatting
   in the PCA.py code.
 - Slightly changed the wording of the Note to improve the accuracy of the
   statement.
 
 Changes to README.txt:
 - Updated the "Running The Codes" section to include a description on running
   if from the directory.
 - Added a note about changing the INFILE variable in the code itself
 ==============================================================================

PCA.py
"""
This code runs a PCA, and plots the results based on the classifications
defined in a Target Column

NOTE: The strings in this code ARE CASE SENSITIVE - make sure you
      match the case for any column names, file names, etc.


Author: Tom Burns
email : tom.burns@canada.ca
date  : Feb 24th 2021

VERSION 1.1.0
"""

VERSION = '%i.%i.%i' % (1, 1, 0)


import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler


# This is the input file that contains your
# descriptors and your target values
#INFILE = 'test.csv'
INFILE = 'BinaryMLFile.csv'
#INFILE = 'TrinaryMLFile.csv'

# This is the heading of the "target" column
# that should contain your classifications
#TARGET = 'Target'
TARGET = 'Class'

# If there are any columns you want the code
# to ignore in the csv file, add them to OMIT
# OMIT = ['Example1', 'Example2']
OMIT = ['CAS']

# If you want to the data to be scaled, set this
# value to True (Recommended),
# otherwise set this value to False
SCALE = True


# ====================================================================
# DO NOT EDIT following lines
LOC = '\\'.join(os.path.realpath(__file__).split('\\')[:-1]) + '\\'
INFILE = LOC + INFILE
# ====================================================================


def import_data(infile=INFILE):
    """Imports and formats the data"""
    # Import the raw data from the file
    raw = pd.read_csv(infile)

    # Filter out the columns to grab only the desired descriptors
    rcols, cols = [c for c in raw], []
    for col in rcols:
        if col not in OMIT and col != TARGET:
            cols.append(col)

    # Format the final arrays
    X, Y = np.array(raw[cols]), np.array(raw[TARGET])
    return X, Y, cols


def plot_results(X, Y, Z, C):
    """Generates the PCA Plots"""
    
    # This defines the colours of each "classification", if you
    # want to add more than 4 classifications, you'll need to add
    # some colours to this list
    clrs = ['b', 'r', 'g', 'y']
    
    # Finds all of the unique classifications
    u = np.unique(Y)
    # Code will error out if length of u > length of clrs
    if len(u) > len(clrs):
        print("Error: Number of unique classifications > Number of defined colours in code")
        exit()
    
    # separate the set into the different classes
    allX, allY = [], []
    for y in u:
        allX.append(X[np.where(Y==y)])
        allY.append(Y[np.where(Y==y)])

    # Plot the final data PCA1 vs PCA2
    plt.subplot(221)
    for i, x in enumerate(allX):
        one, two = x[:,0], x[:,1]
        plt.scatter(one, two, color='none', edgecolor=clrs[i], label='Class %i' % u[i])
    plt.xlabel('PC 1')
    plt.ylabel('PC 2')
    plt.legend()

    # Plot the contribution of each PC to the variance
    plt.subplot(222)
    expl, csum = Z[0], Z[1]
    if len(expl) > 15:
        expl_X = [i+1 for i in range(0, len(expl[:15]))]
        expl_Y = expl[:15]
    else:
        expl_X = [i+1 for i in range(0, len(expl))]
        expl_Y = expl[:]

    plt.title('PC1 + PC2 = %.2f' % csum + '% of the Variance')
    plt.bar(expl_X, 100* expl_Y, edgecolor='b', lw=2, fc=(0, 0, 1, 0.3))
    plt.xlabel('PC')
    plt.ylabel('Contribution to Variance [%]')

    # Descriptor contribution to PC 1
    loadings = Z[2]
    plt.subplot(223)
    first = loadings[0, :]
    first = np.array([abs(i) for i in first])
    dic = pd.DataFrame(first, dtype='float64')
    dic = dic.sort_values([0], ascending=False)
    f_ids = []
    for i in dic.index:
        f_ids.append(i)
    if len(dic) > 15:
        f_y = np.array(dic[0][:15])
        f_x = np.array([i for i in range(len(f_y))][:15])
    else:
        f_y = np.array(dic[0][:])
        f_x = np.array([i for i in range(len(f_y))][:])
    plt.bar(f_x, 100 * f_y, edgecolor='b', lw=2, fc=(0, 0, 1, 0.3))
    plt.ylabel('PC1 Contribution [%]')
    f_labels = [C[i] for i in f_ids]
    plt.xlabel('Variable')
    if len(f_x) < len(f_labels):
        f_labels = f_labels[:len(f_x)]
    plt.xticks(f_x, f_labels, rotation=90)

    # Descriptor Contribution to PC2
    plt.subplot(224)
    first = loadings[1, :]
    first = np.array([abs(i) for i in first])
    dic = pd.DataFrame(first, dtype='float64')
    dic = dic.sort_values([0], ascending=False)
    f_ids = []
    for i in dic.index:
        f_ids.append(i)
    if len(dic) > 15:
        f_y = np.array(dic[0][:15])
        f_x = np.array([i for i in range(len(f_y))][:15])
    else:
        f_y = np.array(dic[0][:])
        f_x = np.array([i for i in range(len(f_y))][:])
    plt.bar(f_x, 100 * f_y, edgecolor='b', lw=2, fc=(0, 0, 1, 0.3))
    plt.ylabel('PC2 Contribution [%]')
    f_labels = [C[i] for i in f_ids]
    plt.xlabel('Variable')
    if len(f_x) < len(f_labels):
        f_labels = f_labels[:len(f_x)]
    plt.xticks(f_x, f_labels, rotation=90)
    #print(sum(100 * f_y))
    #print(f_y)
    plt.show()


def run_pca(X):
    """Runs the PCA fitting"""
    # Loading in the PCA module
    #print(X.shape[1])
    #exit()
    #pca = PCA(n_components=X.shape[1])
    pca = PCA()
    
    # Here is where the descriptor scaling takes place
    if SCALE:
        scaler = StandardScaler()
        sX = scaler.fit_transform(X)
    else:
        sX = X[:]
    
    # Run the PCA
    pcaX = pca.fit_transform(sX)

    # Post Analysis on PCA results
    loadings = pca.components_
    explained = pca.explained_variance_
    explained_ratio = pca.explained_variance_ratio_
    components = np.transpose(pca.components_[0:2, :])
    c_sum = sum(explained_ratio[:2]) * 100

    return pcaX, (explained_ratio, c_sum, loadings)


def welcome():
    """Adds a welcome message"""
    print('=' * 75)
    print(' Starting PCA.py Version %s' % VERSION)
    print('-' * 75)
    print(' Code Parameters:\n')
    print(' Input File       :', INFILE.split('\\')[-1])
    print(' Target Column    :', TARGET)
    print(' Omitting Columns :', OMIT)
    print(' Scale Descriptors:', SCALE)
    print('=' * 75)


def main():
    """Main Execution of Script"""
    welcome()
    X, Y, cols = import_data()
    pcaX, pca_info = run_pca(X)
    plot_results(pcaX, Y, pca_info, cols)


if __name__ in '__main__':
    main()
    print('\n Code Terminated Normally.\n')
    input(' Press enter to close window.\n')


Model_formatter.py
"""
Formats the csv file into a pkl read for model optimization
"""


import os
import numpy as np
import pandas as pd
import pickle as pkl


INFILE = 'BinaryMLFile_edit.csv'
OFILE  = 'BinaryMLFile_edit.pkl'

# ====================================================================
# DO NOT EDIT following lines
LOC = '\\'.join(os.path.realpath(__file__).split('\\')[:-1]) + '\\'
INFILE = LOC + INFILE
OFILE  = LOC + OFILE
# ====================================================================


OMIT = ['CAS']


def main():
    """Main execution of script"""
    data = pd.read_csv(INFILE)
    rcols, cols = [col for col in data], []
    for col in rcols:
        if col in OMIT:
            continue
        cols.append(col)
    frmtd = []
    for i in data.index:
        sub = []
        for col in cols:
            sub.append(data[col][i])
        frmtd.append(sub)
    frmtd = np.array(frmtd)
    pkl.dump(frmtd, open(OFILE, 'wb'))
    print(frmtd)


if __name__ in '__main__':
    main()
    print("\nProgram Terminated Normally.\n")
    input()


DecisionTree.py
"""
Runs a Decision Tree Classifier on the data set returns
analysis of first two nodes in the tree or array of trees.

Goal: to inform the user of the usefulness of each individual
      descriptor during the classification

This code will run several decision tree fittings with a random
set of the descriptors, and return the distributions of those
descriptors in final array of trees. The number of trees run
is determined by the NTREES variable. If only one tree is selected
the code will use all available descriptors in the fitting

NOTE: if the PCA fails to distinguish between any of the descriptors
      the results of these decision trees are not always useful.


Author: Tom Burns
email : tom.burns@canada.ca
date  : Feb 24th 2021

VERSION 1.1.0
"""

VERSION = '%i.%i.%i' % (1, 1, 0)


import os
import sys
import numpy as np
from math import floor
from PCA import import_data
import sklearn.tree as TREE
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.tree import DecisionTreeClassifier as DT


# This is the input file that contains your
# descriptors and your target values
#INFILE = 'test.csv'
INFILE = 'BinaryMLFile.csv'
#INFILE = 'TrinaryMLFile.csv'

# This is the heading of the "target" column
# that should contain your classifications
#TARGET = 'Target'
TARGET = 'Class'

# If there are any columns you want the code
# to ignore in the csv file, add them to OMIT
#OMIT = ['Example1', 'Example2']
OMIT = ['CAS']

# If you want to the data to be scaled, set this
# value to True (Recommended),
# otherwise set this value to False
SCALE = True

# How many trees will you run?
NTREES = 10000

# Run the boostrapping? (Default=False)
BOOT   = False
# If boostrap being run, how many steps
BSTEPS = 10000

# ====================================================================
# DO NOT EDIT following lines
LOC = '\\'.join(os.path.realpath(__file__).split('\\')[:-1]) + '\\'
INFILE = LOC + INFILE
# ====================================================================


def tree_to_code(tree, feature_names):
    """Converts the tree to working code, uses that code to extract the node identities"""
    _tree = TREE._tree
    tree_ = tree.tree_
    feature_name = [
        feature_names[i] if i != _tree.TREE_UNDEFINED else "undefined!"
        for i in tree_.feature
    ]
    #print("def tree({}):".format(", ".join(feature_names)))
    node_dist = {}

    def recurse(node, depth, node_dist):
        #print(depth)
        indent = "  " * depth
        Depth = 'Depth %i' % depth
        if tree_.feature[node] != _tree.TREE_UNDEFINED:
            name = feature_name[node]
            threshold = tree_.threshold[node]
            if Depth not in node_dist:
                node_dist[Depth] = []
            if name not in node_dist[Depth]:
                node_dist[Depth].append(name)
            #print("{}if {} <= {}:".format(indent, name, threshold))
            node_dist = recurse(tree_.children_left[node], depth + 1, node_dist)
            #print("{}else:  # if {} > {}".format(indent, name, threshold))
            node_dist = recurse(tree_.children_right[node], depth + 1, node_dist)
        else:
            #print("{}return {}".format(indent, tree_.value[node]))
            pass
        return node_dist

    node_dist = recurse(0, 1, node_dist)
    return node_dist


def fit_dt(x, y, c, feats):
    """Runs the decision tree fitting"""
    tree = DT(max_depth=2, max_features=feats)
    if SCALE:
        scaler = StandardScaler()
        sx = scaler.fit_transform(x)
    else:
        sx = x[:]
    tree.fit(x, y)
    pred = tree.predict(x)
    ba  = accuracy(pred, np.array(y))
    nodes = tree_to_code(tree, c)
    return nodes, ba


def bootstrap(bset, N=BSTEPS):
    """Runs a boostrapping of the values in bset with N steps"""
    vals, b = [], len(bset)
    print('\n Running Boostrap:')
    for n in range(N):
        sub = []
        for m in range(b):
            sub.append(np.random.choice(bset))
        vals.append(np.mean(sub))
        printProgress(n+1, N)
    print('\nDone.')
    u, s = np.mean(vals), np.std(vals)
    l, h = u - 1.96*s, u + 1.96*s # Ranges for 95% CI
    print('\n' + '~'*75 + '\n Boostrap Results:\n')
    print(' Mean                    = %.2f' % u + ' %')
    print(' Standard Error          = %.2f' % s + ' %')
    print(' 95% Confidence Interval =' + ' %.2f - %.2f' % (l, h) + ' %')
    print('~'*75)
    return u, s, l, h


def accuracy(pred, actu):
    """Calculates the BA"""
    TC, FC, AC = [], [], [] # True Calls, False Calls, All Calls
    UQ = np.unique(actu)
    for i in UQ:
        TC.append(0)
        FC.append(0)
        AC.append(0)
    for i, a in enumerate(actu):
        j = np.where(UQ == a)[0][0]
        p = pred[i]
        if p == a:
            TC[j] += 1
        else:
            k = np.where(UQ == p)[0][0]
            FC[j] += 1
        AC[j] += 1
    BA = []
    for i, v in enumerate(AC):
        if v == 0:
            continue # only a valid skip here because I'm looking at full set.
        BA.append(100 * TC[i] / v)
    return np.mean(BA)


def run_trees(x, y, c):
    """controls the running and analysis of the dt fittings"""
    # These two objects will contain a running tally of the identiies
    # of the first two nodes in the trees
    node_one, node_two = {}, {}

    # How many features will each decision tree consider?
    # If only one tree is being run, it will consider all
    # features, otherwise it will consider N=sqrt(all features),
    # selecting at random.
    if NTREES == 1:
        feats = x.shape[1]
    else:
        feats = floor(x.shape[1]**0.5)

    # Run the actual trees, sort resulting nodes
    print(' Starting Decision Tree Fittings:')
    accs = []
    for run in range(NTREES):
        nodes, ba = fit_dt(x, y, c, feats)
        accs.append(ba)
        for node in nodes['Depth 1']:
            if node not in node_one:
                node_one[node] = 0
            node_one[node] += 1
        for node in nodes['Depth 2']:
            if node not in node_two:
                node_two[node] = 0
            node_two[node] += 1
        printProgress(run+1, NTREES)
    print('\n Done.')
    if BOOT:
        uBA, sBA, lBA, hBA = bootstrap(accs)
    else:
        uBA, sBA, lBA, hBA = None, None, None, None
        print('\n' + '~'*75 + '\n Balanced Accuracy Distribution Statistics:\n')
        print(' Mean               = %.2f' % np.mean(accs) + ' %')
        print(' Standard Deviation = %.2f' % np.std(accs) + ' %')
        print('~'*75)
    return node_one, node_two, (accs, uBA, sBA, lBA, hBA)


def sort_data(iset):
    """Sorts the node counts into usable arrays for plotting"""
    rX, rY = [], []
    for i in iset:
        rX.append(i)
        rY.append(iset[i])
    ri = np.argsort(np.array(rY))
    ri = np.flip(ri)
    X, Y = [], []
    for i in ri:
        X.append(rX[i])
        Y.append(rY[i])
    return np.array(X), np.array(Y)
    

def printProgress (iteration, total, prefix='Progress', suffix='Complete ',
                   decimals=2, barLength=50):
    """
    Call in a loop to create terminal progress bar
    @params:
        iteration   - Required  : current iteration (Int)
        total       - Required  : total iterations (Int)
        prefix      - Optional  : prefix string (Str)
        suffix      - Optional  : suffix string (Str)
    """
    filledLength = int(round(barLength * iteration / float(total)))
    percents = round(100.00 * (iteration / float(total)), decimals)
    bar = '#' * filledLength + '-' * (barLength - filledLength)
    sys.stdout.write(' %s [%s] %.2f%s %s\r' % (prefix, bar, percents, '%', suffix)),
    sys.stdout.flush()


def plot_results(one, two, accs):
    """Plots the distributions from the first two nodes in the DTs"""
    # Format the data from first two nodes
    x_one, y_one = sort_data(one)
    x_two, y_two = sort_data(two)
    
    # The maximum number of values in the x axis of plots 121 and 122
    MAX_X = 20

    # Plot node 1 distribution
    plt.subplot(131)
    plt.bar(x_one[:MAX_X], 100 * y_one[:MAX_X] / NTREES, lw=2, fc=(0, 0, 1, 0.3), edgecolor='b')
    plt.title('First Node Distribution')
    plt.xlabel('Descriptor')
    plt.ylabel('Frequency [%]')
    plt.xticks(x_one[:MAX_X], x_one[:MAX_X], rotation=90)

    # Plot node 2 distribution
    plt.subplot(132)
    plt.bar(x_two[:MAX_X], 100 * y_two[:MAX_X] / NTREES, lw=2, fc=(0, 0, 1, 0.3), edgecolor='b')
    plt.title('Second Node Distribution')
    plt.xlabel('Descriptor')
    plt.ylabel('Frequency [%]')
    plt.xticks(x_two[:MAX_X], x_two[:MAX_X], rotation=90)
    
    # Plot accuracy distribution
    plt.subplot(133)
    plt.hist(accs[0], lw=2, fc=(0,0,1,0.3), edgecolor='b')
    if BOOT:
        plt.axvline(accs[1], color='k', linestyle='-', label='Bootstrap Mean')
        plt.axvline(accs[3], color='k', linestyle=':', label='95% CI')
        plt.axvline(accs[4], color='k', linestyle=':')

    else:
        plt.axvline(np.mean(accs[0]), color='k', linestyle='-', label='Mean')
        plt.axvline(np.mean(accs[0]) - np.std(accs[0]), color='k',
                    linestyle=':', label='Standard Deviation')
        plt.axvline(np.mean(accs[0]) + np.std(accs[0]), color='k', linestyle=':')
    plt.legend()   
    plt.ylabel('Frequency')
    plt.xlabel('Balanced Accuracy [%]')
    plt.title('Balanced Accuracy Distribution')

    plt.show()


def welcome():
    """Adds a welcome message"""
    print('=' * 75)
    print(' Starting DecisionTree.py Version %s' % VERSION)
    print('-' * 75)
    print(' Code Parameters:\n')
    print(' Input File       :', INFILE.split('\\')[-1])
    print(' Target Column    :', TARGET)
    print(' Omitting Columns :', OMIT)
    print(' Scale Descriptors:', SCALE)
    print(' Number of Trees  :', NTREES)
    print(' Run Bootstrap    :', BOOT)
    if BOOT:
        print(' Bootstrap Steps  :', BSTEPS)
    if BOOT and NTREES > 1000:
        print('\n WARNING: You are running a bootstrap on %i Trees,' % NTREES)
        print('          this will be a long calculation and is')
        print('          unlikely to provide useful insights.')
    elif BOOT and NTREES < 8:
        print('\n WARNING: You are running a bootstrap on %i Trees,' % NTREES)
        print('          a minimum of 8 Trees will be required to')
        print('          provide useful insights.')
    print('=' * 75)


def main():
    """Main Execution of Script"""
    welcome()
    X, Y, cols = import_data(infile=INFILE)
    node_one, node_two, accs = run_trees(X, Y, cols)
    plot_results(node_one, node_two, accs)


if __name__ in '__main__':
    main()
    print('\n Code Terminated Normally.\n')
    input(' Press enter to close window.\n')
